### PR TITLE
fix #972: test APITest.testSetterInNodes()

### LINKED ISSUE #972: test APITest.testSetterInNodes() does not work at all
The API contract which should be enforced by the APITest.testSetterInNodes() IS NOT CHECKED, because this test is actually doing NOTHING.

Put the breakpoint on line 359
```java
assertTrue("Check the method " + statement.getParent(CtMethod.class).getSignature() ...
```
and you will see it NEVER enters the loop.

I have made a PR #968, which fixes that bug.... 

Yes Travis reports the errors, but it is not bug in the PR #968, it is report about broken API contracts!

I am not going to fix PR #968, because I have no idea how to do it. But I guess it is good idea to fix it to avoid merging more and more invalid APIs.

### PR DESCRIPTION
**Pull request by surli**
fix #972
Change test condition for testing setter: originally an OR was used where an AND should be used IMHO.

### CODE REVIEW

**Comment by monperrus** [comment]
remove noise + fix bug (if any)

**Comment by surli** [comment]
Add a condition in the test to ignore annotation UnsettableProperties.
Some setter had to be fixed so choices have been made: 
  - in CtFieldImpl and CtLocalVariableImpl an UnsettableProperty has been added to setAssignment as it does not refer to a property but call setDefaultExpression
  - in CtLoopImpl, CtTryImpl, and CtExecutableImpl setBody has been changed to add a conditional test on statement attribute : it's redundant with the test on body, but I was not sure if I needed to change the test to ignore some setters or the method here...
  - in CtVariableAccessImpl I added the type.setParent as expected in the contract of the test, but should it contain one ?

**Comment by monperrus** [comment]
looks almost ready to be merged. @pvojtechovsky you created this issue, OK for you?

File: src/test/java/spoon/test/api/APITest.java [hunk 1]
```diff
@@ -355,11 +371,13 @@ public void accept(CtVisitor visitor) {
 
 		final List<CtMethod<?>> setters = Query.getElements(launcher.getFactory(), new SetterMethodWithoutCollectionsFilter(launcher.getFactory()));
```
Comment (pvojtechovsky) [inline src/test/java/spoon/test/api/APITest.java]: I suggest to check the size of the setters array. At least with the code like
```java
assertTrue(setters.size()>0)
```
actually there is 69 methods and at the time of the bug the collection size was 0.

**Comment by pvojtechovsky** [comment]
I have added two comments. It is up to You if you like it and implement it. The code is OK even without these changes.
Thanks to @surli for fixing!

File: src/main/java/spoon/support/reflect/code/CtVariableAccessImpl.java [hunk 2]
```diff
@@ -58,6 +58,9 @@
 	@Override
 	public <C extends CtTypedElement> C setType(CtTypeReference<T> type) {
 		if (type != null) {
+			type.setParent(this);
```
Comment (surli) [inline src/main/java/spoon/support/reflect/code/CtVariableAccessImpl.java]: Then the contract should not be respected for this specific setter. I could simply ignore it, but I imagine in the future it won't be the only one. 
So should I create a specific annotation to mark setters which does not respect this contract ? Or do you have another idea ?

Comment (monperrus) [inline src/main/java/spoon/support/reflect/code/CtVariableAccessImpl.java]: yes, this is why the getter in the interface is @DerivedProperty.

Comment (surli) [inline src/main/java/spoon/support/reflect/code/CtVariableAccessImpl.java]: Then I can ignore all setters with @DerivedProperty?

**Comment by monperrus** [comment]
2 solutions:

- ignore all setters for which the getter is @DerivedProperty

- put @UnsettableProperty on the setter.

what's the best to your opinon?

**Comment by surli** [comment]
Put @UnsettableProperty on the setter does not look good, as the property is set, but not in the current object. 

The second option seems cleaner, I'll take a look on it.

**Comment by monperrus** [comment]
looks good. ready to merge for you?

**Comment by surli** [comment]
Ok for me.

**Comment by monperrus** [comment]
first merged PR :-) for @surli. welcome to the wonderful world of Spoon!

**Pull request merged**